torch/return_types.py

```diff
@@ -10,8 +10,10 @@
 
 
 def pytree_register_structseq(cls):
-    if not torch.utils._pytree.is_structseq_class(cls):
-        warnings.warn(f"Class {cls!r} is not a PyStructSequence class.")
+    if torch.utils._pytree.is_structseq_class(cls):
```

Comment:
@XuehaiPan: what's the motivation for having all PyStructSequence be pytrees? All namedtuples being pytrees is already unexpected behavior for some people.

Reply:
> what's the motivation for having all PyStructSequence be pytrees?

@zou3519 The idea is there might be a use case that someone writes a custom operator and wants it supported by pytree by default. Currently, if a third-party package has a custom return type, it needs to call `torch.return_types.pytree_register_structseq` manually.

There is also a codegen utility to generate C++ code for PyStructSequence type and assign it to `torch._C._return_types`.

https://github.com/pytorch/pytorch/blob/2a3d8e50fb2923990a5549de63ea38a2e0671cc6/tools/autograd/gen_python_functions.py#L751-L768

See also:

- [torchgen/api/python.py](https://github.com/pytorch/pytorch/blob/main/torchgen/api/python.py)
- [tools/autograd/gen_python_functions.py](https://github.com/pytorch/pytorch/blob/main/tools/autograd/gen_python_functions.py)
- [tools/autograd/templates/python_return_types.h](https://github.com/pytorch/pytorch/blob/main/tools/autograd/templates/python_return_types.h)
- [tools/autograd/templates/python_return_types.h](https://github.com/pytorch/pytorch/blob/main/tools/autograd/templates/python_return_types.h)

------

> All namedtuples being pytrees is already unexpected behavior for some people.

I agree that some people want their namedtuple class to be an opaque object rather than a container. Using `dataclasses.dataclass` might be an alternative for them. Before this PR, all `namedtuple`s are already pytrees and all dataclasses are leaf nodes by default.

See also discussion for making all dataclasses be pytrees by default:

- pytorch/pytorch#106160
- https://github.com/pytorch/pytorch/pull/93214#discussion_r1094096376
- https://github.com/google/jax/issues/2371#issuecomment-640055295

Reply:
> The idea is there might be a use case that someone writes a custom operator and wants it supported by pytree by default. Currently, if a third-party package has a custom return type, it needs to call `torch.return_types.pytree_register_structseq` manually.

What does JAX's pytree do on structseq types?

For custom ops that return structseq: I don't think there is a way to easily do this, so I wouldn't worry about that use case.

Reply:
> What does JAX's pytree do on structseq types?

By default, `structseq` are leaf nodes and all `namedtuple`s are pytree nodes in JAX's pytree. Users need to manually register `structseq` types are pytree nodes like what we do for `pytree_register_structseq`. The JAX community uses `namedtuple`s instead of `structseq`s.

If you search `PyStructSequence` in https://github.com/google/jax and https://github.com/tensorflow/tensorflow, you will get an empty result.

https://github.com/search?q=repo%3Agoogle%2Fjax+PyStructSequence&type=code
https://github.com/search?q=repo%3Atensorflow%2Ftensorflow+PyStructSequence&type=code

Here are lists of `structseq` types:

<details>
<summary>CPython internals</summary>

```python
In [1]: import optree

In [2]: list(filter(optree.is_structseq_class, tuple.__subclasses__()))
Out[2]: 
[
    <class 'sys.int_info'>,
    <class 'sys.float_info'>,
    <class 'UnraisableHookArgs'>,
    <class 'sys.hash_info'>,
    <class 'sys.version_info'>,
    <class 'sys.flags'>,
    <class 'sys.thread_info'>,
    <class 'asyncgen_hooks'>,
    <class '_thread._ExceptHookArgs'>,
    <class 'os.stat_result'>,
    <class 'os.statvfs_result'>,
    <class 'os.terminal_size'>,
    <class 'posix.times_result'>,
    <class 'posix.uname_result'>,
    <class 'time.struct_time'>,
    <class 'resource.struct_rusage'>,
    <class '_lsprof.profiler_entry'>,
    <class '_lsprof.profiler_subentry'>
]
```

</details>

<details>
<summary>With `torch.return_types`</summary>

```python
In [3]: import torch

In [4]: list(filter(optree.is_structseq_class, tuple.__subclasses__()))
Out[4]: 
[
    <class 'sys.int_info'>,
    <class 'sys.float_info'>,
    <class 'UnraisableHookArgs'>,
    <class 'sys.hash_info'>,
    <class 'sys.version_info'>,
    <class 'sys.flags'>,
    <class 'sys.thread_info'>,
    <class 'asyncgen_hooks'>,
    <class '_thread._ExceptHookArgs'>,
    <class 'os.stat_result'>,
    <class 'os.statvfs_result'>,
    <class 'os.terminal_size'>,
    <class 'posix.times_result'>,
    <class 'posix.uname_result'>,
    <class 'time.struct_time'>,
    <class 'resource.struct_rusage'>,
    <class '_lsprof.profiler_entry'>,
    <class '_lsprof.profiler_subentry'>,
    <class 'numpy.core.multiarray.typeinfo'>,
    <class 'numpy.core.multiarray.typeinforanged'>,
    <class 'torch.return_types._fake_quantize_per_tensor_affine_cachemask_tensor_qparams'>,
    <class 'torch.return_types._fused_moving_avg_obs_fq_helper'>,
    <class 'torch.return_types._linalg_det'>,
    <class 'torch.return_types._linalg_det_out'>,
    <class 'torch.return_types._linalg_eigh'>,
    <class 'torch.return_types._linalg_eigh_out'>,
    <class 'torch.return_types._linalg_slogdet'>,
    <class 'torch.return_types._linalg_slogdet_out'>,
    <class 'torch.return_types._linalg_solve_ex'>,
    <class 'torch.return_types._linalg_solve_ex_out'>,
    <class 'torch.return_types._linalg_svd'>,
    <class 'torch.return_types._linalg_svd_out'>,
    <class 'torch.return_types._lu_with_info'>,
    <class 'torch.return_types._scaled_dot_product_efficient_attention'>,
    <class 'torch.return_types._scaled_dot_product_flash_attention'>,
    <class 'torch.return_types._unpack_dual'>,
    <class 'torch.return_types.aminmax'>,
    <class 'torch.return_types.aminmax_out'>,
    <class 'torch.return_types.cummax'>,
    <class 'torch.return_types.cummax_out'>,
    <class 'torch.return_types.cummin'>,
    <class 'torch.return_types.cummin_out'>,
    <class 'torch.return_types.frexp'>,
    <class 'torch.return_types.frexp_out'>,
    <class 'torch.return_types.geqrf_out'>,
    <class 'torch.return_types.geqrf'>,
    <class 'torch.return_types.histogram_out'>,
    <class 'torch.return_types.histogram'>,
    <class 'torch.return_types.histogramdd'>,
    <class 'torch.return_types.kthvalue'>,
    <class 'torch.return_types.kthvalue_out'>,
    <class 'torch.return_types.linalg_cholesky_ex'>,
    <class 'torch.return_types.linalg_cholesky_ex_out'>,
    <class 'torch.return_types.linalg_eig'>,
    <class 'torch.return_types.linalg_eig_out'>,
    <class 'torch.return_types.linalg_eigh'>,
    <class 'torch.return_types.linalg_eigh_out'>,
    <class 'torch.return_types.linalg_inv_ex'>,
    <class 'torch.return_types.linalg_inv_ex_out'>,
    <class 'torch.return_types.linalg_ldl_factor'>,
    <class 'torch.return_types.linalg_ldl_factor_out'>,
    <class 'torch.return_types.linalg_ldl_factor_ex'>,
    <class 'torch.return_types.linalg_ldl_factor_ex_out'>,
    <class 'torch.return_types.linalg_lstsq'>,
    <class 'torch.return_types.linalg_lstsq_out'>,
    <class 'torch.return_types.linalg_lu'>,
    <class 'torch.return_types.linalg_lu_out'>,
    <class 'torch.return_types.linalg_lu_factor'>,
    <class 'torch.return_types.linalg_lu_factor_out'>,
    <class 'torch.return_types.linalg_lu_factor_ex'>,
    <class 'torch.return_types.linalg_lu_factor_ex_out'>,
    <class 'torch.return_types.linalg_qr'>,
    <class 'torch.return_types.linalg_qr_out'>,
    <class 'torch.return_types.linalg_slogdet'>,
    <class 'torch.return_types.linalg_slogdet_out'>,
    <class 'torch.return_types.linalg_solve_ex'>,
    <class 'torch.return_types.linalg_solve_ex_out'>,
    <class 'torch.return_types.linalg_svd'>,
    <class 'torch.return_types.linalg_svd_out'>,
    <class 'torch.return_types.lu_unpack'>,
    <class 'torch.return_types.lu_unpack_out'>,
    <class 'torch.return_types.max'>,
    <class 'torch.return_types.max_out'>,
    <class 'torch.return_types.median'>,
    <class 'torch.return_types.median_out'>,
    <class 'torch.return_types.min'>,
    <class 'torch.return_types.min_out'>,
    <class 'torch.return_types.mode'>,
    <class 'torch.return_types.mode_out'>,
    <class 'torch.return_types.nanmedian'>,
    <class 'torch.return_types.nanmedian_out'>,
    <class 'torch.return_types.qr_out'>,
    <class 'torch.return_types.qr'>,
    <class 'torch.return_types.slogdet'>,
    <class 'torch.return_types.slogdet_out'>,
    <class 'torch.return_types.sort_out'>,
    <class 'torch.return_types.sort'>,
    <class 'torch.return_types.svd_out'>,
    <class 'torch.return_types.svd'>,
    <class 'torch.return_types.topk_out'>,
    <class 'torch.return_types.topk'>,
    <class 'torch.return_types.triangular_solve_out'>,
    <class 'torch.return_types.triangular_solve'>,
    <class 'pwd.struct_passwd'>,
    <class 'grp.struct_group'>
]
```

</details>

Reply:
> IMO we should be consistent, so we should not register structseq as pytree nodes by default. However, users can manually add them (and we "manually" add them for all of the torch.return_types).

@zou3519 Do you mean to be consistent with JAX pytree or the old behavior of pytree in PyTorch?

------

> Does optree support structseq types as pytree nodes by default?

Yes. The motivations are:

1. `PyStructSequence` is a concept in Python C API but shares some similar semantics of `namedtuple`.

	> Python Documentation: [Struct Sequence Objects](https://docs.python.org/3/c-api/tuple.html?highlight=structseq#struct-sequence-objects)
	>> Struct sequence objects are the C equivalent of [namedtuple()](https://docs.python.org/3/library/collections.html#collections.namedtuple) objects, i.e. a sequence whose items can also be accessed through attributes. To create a struct sequence, you first have to create a specific struct sequence type.

	Also, the CPython upstream may want to add members to `PyStructSequence` to make it consistent with `namedtuple` in Python side.
	
	- python/cpython#46145
	- python/cpython#108648

	Since we treat all `namedtuple` types are pytree nodes, it is reasonable to have all PyStructSequence types be pytree nodes if the upstream PR is landed.
	
	Here is the implementation for `is_namedtuple_class` in the previous PR in the ghstack:
	
	```python
	def is_namedtuple_class(cls):
	    return (
	        isinstance(cls, type)
	        and issubclass(cls, tuple)
	        and isinstance(getattr(cls, "_fields", None), tuple)
	        and all(type(field) is str for field in cls._fields)
	        and callable(getattr(cls, "_make", None))    # <= structseq fails here even if python/cpython#108648 landed
	        and callable(getattr(cls, "_asdict", None))
	    )
	```
	
	We will get `is_namedtuple(structseq)` to be `False` even if python/cpython#108648 landed.
	
	But for JAX pytree, it only checks the presence of `_fields`:
	
	```python
	def is_namedtuple(obj):
	    return isinstance(obj, tuple) and hasattr(obj, '_fields')
	```
	
	All PyStructSequence instances will be namedtuple if python/cpython#108648 landed.

2. `PyStructSequence` types are subclasses of `tuple`, and they are container types. This fits the definition of pytree node. That's why we manually register all types in `torch.return_types.*` as pytree node. Here is an example usage that requires this: `tree_map(lambda x: x.cuda(), tree)`.

3. The JAX community does not use `PyStructSequence`.
4. The number of `PyStructSequence` is limited (see https://github.com/pytorch/pytorch/pull/113258#discussion_r1411157204). Because it is related to C code and there is no easy way to create a new type from the Python side.

------

> If so, we probably want some way to toggle this behavior to make it consistent.

I can add an API to have namespace-specific configurations in optree. But before that, I want to hear your thoughts about the explanations above.

Reply:
> > All PyStructSequence instances will be namedtuple if landed
> I would wait until those PRs actually land before changing the behavior, to be consistent

Should we ask on the PR when they think they'll merge it? lol

Reply:
> Should we ask on the PR when they think they'll merge it? lol

Based on the label of the issue, the initial plan is to have this in Python 3.9 but delayed to 3.13. I think they will merge relevant changes eventually.

- python/cpython#46145
- python/cpython#46561
- python/cpython#108648